src/szk.rs

```diff
@@ -0,0 +1,444 @@
+                return Ok(false);
+            };
+        };
+
+        Ok(true)
+    }
+}
+
+mod tests {
+    use super::*;
+    use crate::field::Field128 as TestField;
+    use crate::flp::types::Sum;
+    use rand::{thread_rng, Rng};
+
+    #[test]
+    fn test_sum() {
+        let sum = Sum::new(11).unwrap();
+        let encoded_measurement = sum.encode_measurement(&9).unwrap();
+        let mut nonce = [0u8; 16];
+        let mut verify_key = [0u8; 16];
+        let algorithm_id = 5;
+        let szk_sum = Szk::new_turboshake128(sum, algorithm_id);
+        let mut coins = [0u8; 48];
+        thread_rng().fill(&mut verify_key[..]);
+        thread_rng().fill(&mut nonce[..]);
+        thread_rng().fill(&mut coins[..]);
+        let mut random = coins.chunks_exact(16);
+        let prove_rand_seed = Seed::<16>::from_bytes(
+            random.next().unwrap()[..]
+                .try_into()
+                .expect("wrong seed size"),
+        );
+        let helper_seed = Seed::<16>::from_bytes(
+            random.next().unwrap()[..]
+                .try_into()
+                .expect("wrong seed size"),
+        );
+        let leader_seed = Seed::<16>::from_bytes(
+            random.next().unwrap()[..]
+                .try_into()
+                .expect("wrong seed size"),
+        );
+        let helper_input_share: Vec<TestField> =
+            Prng::from_seed_stream(XofTurboShake128::seed_stream(&helper_seed, &[2u8; 8], &[0]))
+                .take(szk_sum.typ.input_len())
+                .collect();
+        let mut leader_input_share = encoded_measurement.clone();
+        for (x, y) in leader_input_share.iter_mut().zip(&helper_input_share) {
+            *x -= *y;
+        }
+
+        let [l_proof_share, h_proof_share] = szk_sum
+            .prove(
+                &leader_input_share,
+                &helper_input_share,
+                &encoded_measurement,
+                [prove_rand_seed, helper_seed],
+                Some(leader_seed),
+                &nonce,
+            )
+            .unwrap();
+
+        let l_query_share = szk_sum
+            .query(&leader_input_share, l_proof_share, &verify_key, &nonce)
+            .unwrap();
+
+        let h_query_share = szk_sum
+            .query(&helper_input_share, h_proof_share, &verify_key, &nonce)
+            .unwrap();
+
+        // Note to self: verifier destroyed here
+        let mut verifier = l_query_share.verifier;
+
+        for (x, y) in verifier.iter_mut().zip(h_query_share.verifier) {
+            *x += y;
+        }
+        let (h_jr_seed, h_joint_rand) = match h_query_share.joint_rand_seed_and_part {
+            Some([jr_seed, jr_part]) => (Some(jr_seed), Some(jr_part)),
+            None => (None, None),
+        };
+        let (l_jr_seed, l_joint_rand) = match l_query_share.joint_rand_seed_and_part {
+            Some([jr_seed, jr_part]) => (Some(jr_seed), Some(jr_part)),
+            None => (None, None),
+        };
```

Comment:
Hmm, this work was already done `query()` correct?

An API that will be more useful for Mastic is one that returns the "expected" joint randomness seed along side the query share. In this test we would want to make sure that the aggregators compute the same expected seed; then we would pass this to `decide()`.

Reply:
I'm not quite understanding. The only thing this block of code is doing is converting the expected  jr seed and jr part from one option as they're stored in the query share (c.f. Tim's comments) into two options so that we can "send" the jr part between leader and helper as a prep message without the jr seed. The mastic query API does already return the expected seed; but decide is performing the test you describe since the aggregator doesn't have the confirmed jr part from the other aggregator until that point.

Reply:
Hmm, perhaps I'm misunderstanding how the protocol is supposed to work. I believe it's this (from the perspective of the Leader):

1. Leader computes its verifier share, the Leader joint rand part, and the joint rand seed. It needs to confirm the joint rand seed is correct. 
2. The Helper sends its verifier share and the Helper joint rand part.
3. The Leader checks that joint rand seed is correct using the Leader's and Helper's joint rand parts.
4. The Leader checks the verifier.

Is this correct?

If so, then my point is: at some point `query()` needs to admit the joint rand seed it used to compute the verifier share, correct?

Reply:
That's correct, yes. Step 1 is the Leader's run of query(); it returns a query share containing its verifier share, the joint rand part, and the joint rand seed. These will need to be extracted separately because the first two (vf and jr part) will get sent to the Helper and the last one (jr seed) will get cached and passed to decide() along with the Helper's messages. That's what's happening in the tagged code. Step 2 is the Helper's run of query(), which does the same thing. 

The Step 3 check that the joint rand seed is correct happens in decide(), and it is based on both the Leader and Helper's jr parts. Step 4 also happens in decide(), although I usually do it before step 3 (the order should not matter). 

If you're asking whether query() needs to take the joint rand seed as input, the answer is no, because it derives the jr seed for the first time from the jr part it is given in the input share. If you're asking whether it needs to return the jr seed, it already does. Are you suggesting that query() needs to return the jr seed as part of a tuple instead of a query_share struct? I think that was another Clippy-hates-complex-data-types decision.